### PR TITLE
Cypress tests for data explorer

### PR DESCRIPTION
**Pull request by abbyhu2000**
### Description
This PR migrates the old selenium tests for discover to cypress tests for data explorer.
PR [for #5003](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5072) needs to be merged first.

Below are a list of bugs found when writing the tests, the tests for the below feature are skipped and commented with the issue link:

- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5071
- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5068
- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5058
- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5057
- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5056

### Issues Resolved
resolves https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4781

### Check List

- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: cypress.json
Comment (kavilla) [inline cypress.json:23]: revert please

File: cypress/fixtures/dashboard/opensearch_dashboards/data_explorer/date_nanos_mix/data.json.txt
Comment (kavilla) [inline cypress/fixtures/dashboard/opensearch_dashboards/data_explorer/date_nanos_mix/data.json.txt:2]: can some of this be zipped up?

File: cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/date_nanos.spec.js
Comment (kavilla) [inline cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/date_nanos.spec.js:33]: do you think it's low hanging fruit enough to make this configurable to where I can pass a cypress env variable if I want legacy discover or current discover?

**Comment by kavilla** [comment]
AWESOME! Will verify later this week. Made some quick comments we should also verify with security enabled.

**Comment by abbyhu2000** [comment]
> AWESOME! Will verify later this week. Made some quick comments we should also verify with security enabled. Targetting 2.11?

yea i think it is targeting 2.11 @ashwin-pc

File: cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/date_nanos.spec.js
Comment (ashwin-pc) [reply cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/date_nanos.spec.js:33]: LEgacy discover will be removed in the next release, so it does not make sense to switch between the two

File: cypress/utils/dashboards/data_explorer/constants.js
Comment (ashwin-pc) [inline cypress/utils/dashboards/data_explorer/constants.js:12]: nit: Stick to `DE`
```suggestion
export const DE_DEFAULT_START_TIME = 'Sep 19, 2015 @ 06:31:44.000';
export const DE_DEFAULT_END_TIME = 'Sep 23, 2015 @ 18:31:44.000';
```

File: cypress/utils/dashboards/data_explorer/commands.js
Comment (ashwin-pc) [inline cypress/utils/dashboards/data_explorer/commands.js:75]: Id remove these two commands and move the code to be next to the test itself for readability. They both appear to be used in just one test.

File: package.json
Comment (ashwin-pc) [inline package.json:44]: Is this necessary?

File: cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/sidebar.spec.js
Comment (ashwin-pc) [inline cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/sidebar.spec.js:12]: Reset to default after the test

**Review by ashwin-pc — COMMENTED** [review]
Firstly, really good job with the PR. The tests look good and well written. Im partway though the review by have a general question about cleanup. You are adding a lot of fixtures and changing settings. @kavilla Do you think we need to remove the fixtures after each test/test suite?

File: cypress/utils/dashboards/data_explorer/constants.js
Comment (abbyhu2000) [reply cypress/utils/dashboards/data_explorer/constants.js]: fixed

File: cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/date_nanos.spec.js
Comment (abbyhu2000) [inline cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/date_nanos.spec.js:40]: for the fixture clean up, we can do something like this for each test file? @kavilla  @ashwin-pc

File: cypress/utils/dashboards/data_explorer/commands.js
Comment (abbyhu2000) [reply cypress/utils/dashboards/data_explorer/commands.js:75]: done

File: package.json
Comment (abbyhu2000) [reply package.json:44]: i think i did this accidentally when i was setting up cypress. It is not necessary

File: cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/sidebar.spec.js
Comment (abbyhu2000) [reply cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/sidebar.spec.js:12]: the default is `null`, do we have to set it back to `null`?

Comment (ashwin-pc) [reply cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/sidebar.spec.js:12]: Oh then no worries. But in that case why is this needed?

Comment (ashwin-pc) [reply cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/sidebar.spec.js:12]: Oh then nvm :)

File: package.json
Comment (ashwin-pc) [reply package.json:44]: Can you remove it then?

**Comment by kavilla** [comment]
> You are adding a lot of fixtures and changing settings. @kavilla Do you think we need to remove the fixtures after each test/test suite?

@ashwin-pc that's a good point, the selenium tests would check if load an add them if needed where as some plugins really rely on a fresh state.

Perhaps instead of creating a helper function to check and optimize it can be apart of the cypress env variables where it defaults deletes these values after each spec. and we can configure it from our CI to not delete per spec.

Another work around (which isn't a catchall) but we ensure the complete uniqueness of the data index created.

**Comment by ashwin-pc** [comment]
P.S. Dont merge this until `feature/deangular` is merged into `main` for OSD.

**Review by ashwin-pc — COMMENTED** [review]
@abbyhu2000 can you add the recordings for these tests to the PR?

File: cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/shared_links.spec.js
Comment (kavilla) [inline cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/shared_links.spec.js:28]: haven't ran this one yet. this doesn't cause a failure when it includes an extra base path right?

Comment (ashwin-pc) [reply cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/shared_links.spec.js:28]: Ive fixed that issue in 2.10, so no :)

File: cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/sidebar.spec.js
Comment (kavilla) [reply cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/data_explorer/sidebar.spec.js:12]: > Reset to default after the test

+1, yes. The value will be null when a user unsets a setting that is default undefined (or null)

**Comment by ashwin-pc** [comment]
@kavilla hmm, that seems like that can come as a followup. Want to put some more thought into that. I'm okay with this going in as is for now and following up with improvements to test improvements later when we have a better solution for cleanup.

**Comment by kavilla** [comment]
@abbyhu2000 @ashwin-pc I think maybe we should consider optimizing the uploading data. Or we start building in the ciGroup stuff since I see this PR increasing the time of the test run:

https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5134#issuecomment-1738307120.

I thought it was failing but I believe it was working just fine just taking a while.

File: package.json
Comment (abbyhu2000) [reply package.json:44]: removed

**Comment by abbyhu2000** [comment]
Wrote a function `importJSONDocIfNeeded` to optimize uploading data, it will save time if the indices are already loaded in a previous test.

**Comment by abbyhu2000** [comment]
Test run on Github CI succeed: https://github.com/opensearch-project/OpenSearch-Dashboards/actions/runs/6427043272/job/17453245212

**Comment by abbyhu2000** [comment]
Success run on the CI: https://github.com/opensearch-project/OpenSearch-Dashboards/actions/runs/6462297926

**Review by kavilla — APPROVED** [review]

**Review by ananzh — APPROVED** [review]

**Pull request merged**
